### PR TITLE
Table reader

### PR DESCRIPTION
**Pull request by Uggah**

### CODE REVIEW

File: src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/databasemodel/Table.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import de.hdm_stuttgart.mi.dbad.dbwarp.databasemodel.constraints.Constraint;
+import java.util.ArrayList;
+import java.util.List;
+import lombok.Data;
+
+@Data
+public class Table {
+
+  final private String name;
+  private ConnectionManager connectionManager;
```
Comment (Uggah) [inline src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/databasemodel/Table.java]: Why does a table need a connection manager?

File: src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/databasemodel/Table.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  private List<Column> columns = new ArrayList<>();
+  private Constraint[] constraints;
```
Comment (Uggah) [inline src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/databasemodel/Table.java]: Why not use a List?

File: src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/databasemodel/Table.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import lombok.Data;
+
+@Data
+public class Table {
+
+  final private String name;
+  private ConnectionManager connectionManager;
+  private List<Column> columns = new ArrayList<>();
```
Comment (Uggah) [inline src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/databasemodel/Table.java]: Can't this be final?

File: src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/migration/MigrationManager.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import de.hdm_stuttgart.mi.dbad.dbwarp.migration.tablereader.TableReader;
+import de.hdm_stuttgart.mi.dbad.dbwarp.migration.tablereader.TableReaderFactory;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+import lombok.SneakyThrows;
+
+@NoArgsConstructor
+public final class MigrationManager {
+
+  private static MigrationManager INSTANCE;
```
Comment (Uggah) [inline src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/migration/MigrationManager.java]: Final, please.

File: src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/migration/tablereader/TableReader.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+  private ConnectionManager connectionManager;
+
+  public TableReader(ConnectionManager connectionManager) {
+    this.connectionManager = connectionManager;
+  }
+
+  ;
+
+  public abstract Table[] readTables() throws SQLException;
```
Comment (Uggah) [inline src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/migration/tablereader/TableReader.java]: Why not return a List<Table>, Set<Table> or Collection<Table>?

File: src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/migration/tablereader/TableReaderFactory.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+import de.hdm_stuttgart.mi.dbad.dbwarp.connection.ConnectionManager;
+import java.sql.SQLException;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@NoArgsConstructor
+public final class TableReaderFactory {
+
+  private static TableReaderFactory INSTANCE;
```
Comment (Uggah) [inline src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/migration/tablereader/TableReaderFactory.java]: Final, please.

File: src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/migration/tablereader/TableReaderFactory.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
+package de.hdm_stuttgart.mi.dbad.dbwarp.migration.tablereader;
+
+import de.hdm_stuttgart.mi.dbad.dbwarp.connection.ConnectionManager;
+import java.sql.SQLException;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@NoArgsConstructor
```
Comment (Uggah) [inline src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/migration/tablereader/TableReaderFactory.java]: `@NoArgsConstructor(accessLevel = AccessLevel.PRIVATE)`

File: src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/migration/tablereader/TableReaderFactory.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+    System.out.println(
+        connectionManager.getSourceDatabaseConnection().getMetaData().getDatabaseProductName());
+
+    switch (connectionManager.getSourceDatabaseConnection().getMetaData()
+        .getDatabaseProductName()) {
+      case "SQLite":
+        return new SQLiteTableReader(connectionManager);
+      default:
+        return null;
```
Comment (Uggah) [inline src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/migration/tablereader/TableReaderFactory.java]: Shouldn't we throw an Exception here?

File: src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/migration/tablereader/TableReaderFactory.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    if (INSTANCE == null) {
+      INSTANCE = new TableReaderFactory();
+    }
+
+    return INSTANCE;
+  }
+
+  public TableReader getTableReader() throws SQLException {
+
+    System.out.println(
```
Comment (Uggah) [inline src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/migration/tablereader/TableReaderFactory.java]: Please use logging...

File: src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/databasemodel/Table.java (hunk 1 above)
Comment (Joystick01) [reply src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/databasemodel/Table.java]: is removed

File: src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/databasemodel/Table.java (hunk 2 above)
Comment (Joystick01) [reply src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/databasemodel/Table.java]: i thought an array might be best practice due to memory efficency

is changed

File: src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/databasemodel/Table.java (hunk 3 above)
Comment (Joystick01) [reply src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/databasemodel/Table.java]: is changed

File: src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/migration/MigrationManager.java (hunk 4 above)
Comment (Joystick01) [reply src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/migration/MigrationManager.java]: not possible

File: src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/migration/tablereader/TableReader.java (hunk 5 above)
Comment (Joystick01) [reply src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/migration/tablereader/TableReader.java]: i thought an array might be best practice due to memory efficency

is changed

File: src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/migration/tablereader/TableReaderFactory.java (hunk 6 above)
Comment (Joystick01) [reply src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/migration/tablereader/TableReaderFactory.java]: not possible

File: src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/migration/tablereader/TableReaderFactory.java (hunk 7 above)
Comment (Joystick01) [reply src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/migration/tablereader/TableReaderFactory.java]: is changed

File: src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/migration/tablereader/TableReaderFactory.java (hunk 8 above)
Comment (Joystick01) [reply src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/migration/tablereader/TableReaderFactory.java]: is changed

File: src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/migration/tablereader/TableReaderFactory.java (hunk 9 above)
Comment (Joystick01) [reply src/main/java/de/hdm_stuttgart/mi/dbad/dbwarp/migration/tablereader/TableReaderFactory.java]: is removed, was a debugging artefact

**Review by Joystick01 — APPROVED** [review]

**Pull request merged**